lib/hal/audio.c

```diff
@@ -55,25 +61,58 @@ static void __stdcall DPC(PKDPC Dpc,
 // PCM actions, since S/PDIF is always spooling through the same buffer.
 static BOOLEAN __stdcall ISR(PKINTERRUPT Interrupt, PVOID ServiceContext)
 {
-	// Was the interrupt triggered because we were out of data?
-	if ((*((char *)0xFEC00116))&8)
-		KeInsertQueueDpc(&DPCObject,NULL,NULL); //calls user callback soon
+	AC97_DEVICE *pac97device = &ac97Device;
+	volatile unsigned char *pb = (unsigned char *)pac97device->mmio;
 
-	//KeInsertQueueDpc queues Dpc and returns TRUE if Dpc not already queued.
-	//Dpc will be queued only once. So only one Dpc is fired after ISRs cease fire.
-	//DPCs avoid crashes inside non reentrant user callbacks called by nested ISRs.
-	//CAUTION : if you use fpu in DPC you have to save & restore yourself fpu state!!!
-	//(fpu=floating point unit, i.e the coprocessor executing floating point opcodes)
+	unsigned char analogInterrupt = pb[0x116];
+	unsigned char digitalInterrupt = pb[0x176];
+
+	bool waitingForAnalog = (analogBufferCount > digitalBufferCount);
+	bool waitingForDigital = (digitalBufferCount > analogBufferCount);
```

Comment:
This must be sampled later, otherwise if we aren't fast enough at handling an interrupt, then we might have to handle 2 interrupts at the same time.
In that situation, none of them would be waiting (because the counts are the same), but both of them would decrement the buffercount (without marking `waitCompleted` which would cause a buffer to be freed).

So to avoid this, this has to be done when the `analogBufferCount` has already been decremented (thereby causing `waitingForDigital` to be `true`, thereby causing `waitCompleted` to be `true`).